iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -321,18 +321,138 @@ struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
   }
 };
 
+/// Converts matrix-times-matrix-transposed vector.contracts with
+/// lhs and rhs inputs defined by arith.extsi promoting from i8 to i32,
+///
+///     %lhs_i32 = arith.extsi %lhs_i8 : i8 to i32
+///     %rhs_i32 = arith.extsi %rhs_i8 : i8 to i32
+///     %result = vector.contract [...]
+///                 %lhs_i32 : vector<8x4xi32>,
+///                 %rhs_i32 : vector<8x4xi32>,
+///                 %acc_i32 : vector<8x8xi32>,
+///                 [...]
+///
+/// To vector ops reading directly from the %lhs_i8 and %rhs_i8 values
+/// (bypassing the existing arith.extsi) and passing that to a llvm.inline_asm
+/// block implementing the matrix multiplication arithmetic using Aarch64
+/// dot-product instructions (sdot).
+/// It matches the same patterns as MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm
+struct MMT_8x4x8_i8i8i32_Aarch64Dotprod_Intrinsics
+    : public OpRewritePattern<vector::ContractionOp> {
+ public:
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractionOp,
+                                PatternRewriter &rewriter) const override {
+    // It would be nice to sharethe matching logic between this and
+    // MMT_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm, but it defines a bunch of
+    // variables, so doesn't end up being that helpful.
```

Comment:
Slightly confused by this comment! IIUC the comment means: "since we would want the local variables defined by this code anyway, it isn't that helpful to try to share that code with the other function"?  Anyway, since you use the same helper functions isMatrixTimesMatrixTransposedOfGivenShape and getExtSIInput, you are already benefiting from the bulk of the sharing benefits, so totally up to you.

Reply:
Yes that's what I mean, and I agree that most of the sharing benefit has been gained. I started trying to factor out this block and it ended up looking nasty, so I stopped.